Editor: Implement meta as custom source

aduth opened the pull request:
Closes #16282
Related (alternative to): #16075

This pull request seeks to explore an implementation of custom sources.

**Implementation Notes:**

The implementation inherits a fair bit from #16075 (notably "last changes" tracking), but it tries to leverage a flow where sourced attributes are updated (as applicable) and applied prior to any blocks reset from the editor module. Furthermore, per discussion at https://github.com/WordPress/gutenberg/pull/16075#discussion_r298365062, they are implemented as standard store controls, where the benefit of this approach is in increased flexibility of the implementation by avoiding the need to determine a specific set of arguments which would apply for all potential custom source implementations. Instead, each implementation can simply yield controls (such as `select` from `@wordpress/data-controls`) to retrieve whichever data they depend upon.

It diverges from some alternative proposals (https://github.com/WordPress/gutenberg/issues/16282#issuecomment-507423808) due to:

- Not having well-defined "start" and "end" cycles from within the context of a block editor
- The difficulty and overhead in keeping redundant data in sync for "sourced" data, where this implementation can derive directly from post meta (rather than to carve a separate state which needs to be kept in sync with the post)
- In the alternative, not being able to eliminate selector-time source derivation (as of the changes here, `getBlockAttributes` returns a static reference)

As to the proposed "Block Sources API", I'm not very attached to the specific set of arguments currently passed to `apply`, `applyAll`, and `update`. The idea for `applyAll` came about as a result of [a performance concern](https://github.com/WordPress/gutenberg/pull/16075#discussion_r298489530) for repeated data access, but it is not strictly necessary (`apply` and `update` alone would be nicely complementing, though similarly we could explore to add an `updateAll`). I didn't seek to make this publicly-extensible for the moment, though it is designed to allow for it in the future after some internal validation of the API.

**Remaining Tasks:**

- [ ] Unit tests to be written, if implementation deemed acceptable.

**Testing Instructions:**

Verify that updating a block with an attribute sourced by a meta attribute reflects the update (it is restored after a save, and it applies to all other blocks which source from the same meta).

As an example, I think this demo plugin should still work: https://gist.github.com/pento/19b35d621709042fc899e394a9387a54

Caveats:

- Block validation fails if the meta value is persisted into post content (#4989). I aim to tackle this separately, where validation occurs as a separated process from the default blocks `parse` behavior and runs only after the meta values are applied.
- There seems to be an issue where other blocks sourced from the same meta do not re-render immediately. Clicking to select those other blocks will trigger the render and reflect the update. I suspect there may be an issue with either the editor's async rendering, or how we choose to memoize these blocks' components rendering.

aduth committed:
Editor: Implement meta as custom source

Co-Authored-By: Riad Benguella <benguella@gmail.com>

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -730,7 +751,20 @@ export function unlockPostSaving( lockName ) {
  */
-export function resetEditorBlocks( blocks, options = {} ) {
+export function* resetEditorBlocks( blocks, options = {} ) {
+	for ( const name in sources ) {
```
I'm not sure if it's an issue with Babel transpilation or a distinct behavior for how named exports "objects" are iterated, but I would have preferred to use a more concise `for ( const source of sources ) {` syntax here, yet I was encountering runtime errors ¯\\_(ツ)\_/¯

epiqueras replied:
`Objects` are not `Iterable`. You would need to `for of Object.keys/values/entries/etc(sources)` or export an array of sources.

epiqueras replied:
This would be a good place to call an `updateAll` if it exists.

aduth replied:
> `Objects` are not `Iterable`. You would need to `for of Object.keys/values/entries/etc(sources)` or export an array of sources.

Oh! That seems obvious in retrospect. I guess `for ( const source of Object.values( sources ) ) ) {` might be what I'm looking for here then.

youknowriad commented on packages/block-editor/src/components/provider/index.js:
```diff
@@ -90,9 +96,19 @@ class BlockEditorProvider extends Component {
+
+				if ( onBlockAttributesChange ) {
+					const [ clientId, attributes ] = newLastBlockAttributesChange;
+					onBlockAttributesChange( clientId, attributes );
```
Very first thought :) Seeing this made me wonder about an idea. Not sure yet how valuable it is or if it will allow us to improve things. but this callback could serve as a way to make the updates the blocks and "return" them.

I didn't read the PR completely but If I'm not mistaken this callback forces the caller to "set blocks" twice. When this callback is called and when `onInput` is called too?

(I might be completely wrong here, as I didn't dive yet)

aduth replied:
> I didn't read the PR completely but If I'm not mistaken this callback forces the caller to "set blocks" twice. When this callback is called and when `onInput` is called too?

No, because `onBlockAttributesChange` isn't actually applying anything to the blocks, it's only triggering the side effect (updating the post). As you mention, it's the subsequent `onChange` / `onInput` which calls `resetBlocks`, at which point those values are reflected in all impacted blocks.

aduth replied:
> No, because `onBlockAttributesChange` isn't actually applying anything to the blocks, it's only triggering the side effect (updating the post). As you mention, it's the subsequent `onChange` / `onInput` which calls `resetBlocks`, at which point those values are reflected in all impacted blocks.

An unfortunate consequence of this behavior: Since we rely on the subsequent `resetBlocks` and there's nothing about `editPost` itself which immediately triggers an application of sourced attribute values, if someone were to call `editPost` directly with updated meta values, the current implementation would not immediately update blocks which derive from this meta.

epiqueras replied:
>An unfortunate consequence of this behavior: Since we rely on the subsequent resetBlocks and there's nothing about editPost itself which immediately triggers an application of sourced attribute values, if someone were to call editPost directly with updated meta values, the current implementation would not immediately update blocks which derive from this meta.

We need a way for custom sources to subscribe to events/actions that should re-`apply` their values.

youknowriad commented on packages/block-editor/src/components/provider/README.md:
```diff
@@ -0,0 +1,41 @@
+BlockEditorProvider
```
❤️ I love this README :)

aduth commented:
>Seeing this made me wonder about an idea. Not sure yet how valuable it is or if it will allow us to improve things. but this callback could serve as a way to make the updates the blocks and "return" them.

This prompted me to consider that we likely have a problem here with how the BlockEditorProvider considers its value as canonical, and upon an "outbound" sync will ignore the value provided in the next render. This could perhaps explain the issues with other meta blocks not being updated immediately. Your idea here could work at least for the one block being updated, but doesn't account for other meta blocks which source from the same meta property.

In another of my experimental branches, I had considered whether it would be enough that the BlockEditorProvider [ensure the next value it receives in a sync matches](https://github.com/WordPress/gutenberg/commit/5adb8ab8b62a7be9c5aa02f54f7e6b65317376bf#diff-f8ab9759c8d63240841c9d18565a9402L37) what it had expected based on what was sent. We could use this in the implementation here, then only "apply" values to blocks if they differ from what was sent by the block editor, thus triggering the BlockEditorProvider to reset its blocks based on what was applied from the editor.

Admittedly, this is one thing the [alternative considered proposal](https://github.com/WordPress/gutenberg/issues/16282#issuecomment-507423808) might handle well, at least so far as the block editor doesn't need to become _aware_ of this new value resulting from a sync, since it retains the role of being the canonical source of values.

aduth commented:
In thinking again about the "Source API" here, and partly with regard to my previous comment at least so far as the added complexity in avoiding to mutate (and only create new references) for blocks "applied", I wonder if we could eliminate `applyAll`, and align what's proposed here as `align` closer to the `update` interface.

In other words:

- `update( attributeSchema, value )`
- `apply( attributeSchema )`

This way, the framework can decide whether the value from `apply` differs from what's expected, cloning the blocks array as necessary.

The downside here is that we revert back to a previously-mentioned [potential performance concern](https://github.com/WordPress/gutenberg/pull/16075#discussion_r298489530). I think this can be mitigated by the fact we only run `apply` when known to be a block of a given source. Furthermore, this only happens at specific intervals (block updates) that it's likely not to become a bottleneck.

aduth commented:
From https://github.com/WordPress/gutenberg/pull/16402#discussion_r300015701:

> An unfortunate consequence of this behavior: Since we rely on the subsequent `resetBlocks` and there's nothing about `editPost` itself which immediately triggers an application of sourced attribute values, if someone were to call `editPost` directly with updated meta values, the current implementation would not immediately update blocks which derive from this meta.

We could solve this in a similar way to #16075 by having built-in awareness to `editPost` that sourced attributes need to be updated, but this breaks down the isolation of a source implementation, lessening the usefulness of this as a generic or reusable interface.

Considering how it might be made to be generic, the sources would need some way to subscribe to the store, and in the case of meta, receive dependent data upon whose changing should cause a re-application of sourced data on blocks.

epiqueras commented on packages/editor/src/store/actions.js:
```diff
@@ -67,8 +68,28 @@ export function* setupEditor( post, edits, template ) {
+
+	for ( const [ attributeName, schema ] of Object.entries( blockType.attributes ) ) {
+		if ( attributes.hasOwnProperty( attributeName ) && sources[ schema.source ] ) {
+			yield* sources[ schema.source ].update( schema, attributes[ attributeName ] );
```
Nice use of generator delegation!

epiqueras commented on packages/editor/src/store/block-sources/README.md:
```diff
@@ -0,0 +1,22 @@
+Block Sources
```
This is awesome! 🚀

epiqueras reviewed (commented):
I am really liking this approach. Props on coming up with such a clean way to introduce it. 👏 

I think the only unsolved problem is:

>We need a way for custom sources to subscribe to events/actions that should re-apply their values.

So that things like direct `editPost` calls don't throw things out of sync. Maybe custom sources can export something like:

```js
export const reApplyOn = [ 'EDIT_POST' ]
```

aduth commented on packages/block-editor/src/components/provider/README.md:
```diff
@@ -0,0 +1,41 @@
+* **Type:** `Function`
+* **Required** `no`
+
+A callback invoked when the blocks have been modified in a non-persistent manner. Contrasted with `onChange`, a "non-persistent" change is one which is part of a composed input. Any sequence of updates to the same block attribute are treated as non-persistent, except for the first.
```
> Why is this desired? If I edit a color twice, the second edit wouldn't persist?

I think the choice of "persistent" as the term here can be potentially misleading. The intent was for it to be the distinction between `input` and `change` events in the DOM API:

>The input event is fired every time the value of the element changes. **This is unlike the change event, which only fires when the value is committed**, such as by pressing the enter key, selecting a value from a list of options, and the like.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event

In the context of the editor, the use-case is for Undo levels. You don't want Cmd+Z to undo paragraph changes one character at a time, but rather as units (in our case, reflected as sequences of updates to the same block attribute).

When I was first thinking about this documentation, I had in mind to explicitly mention how it's used for Undo/Redo in the editor, but neglected to write it when I'd returned to my computer 😄

epiqueras replied:
Thanks, that's much clearer now. I remember running into that code, but forgot about it when reading this.

aduth commented on packages/block-editor/src/components/provider/README.md:
```diff
@@ -0,0 +1,41 @@
+### `value`
+
+* **Type:** `Array`
+* **Required** `no`
```
> The component will try to use all of these props (except for "children") and throw errors. We should mark them as required.

I was actually thinking we should make them optional. At least in the case of `onInput` and `onChange`, it's quite likely most usage will only care to provide one or the other.

I agree that the documentation here is not accurate per the current implementation. I didn't want to document an undesirable requirement, however. Maybe we should seek to make it optional as a separate task to this pull request.

epiqueras replied:
Makes sense 👍

aduth commented:
>Can we make this part of the onChange/onInput actions instead of adding another action?

Maybe. @youknowriad mentioned a specific concern to me that this might help address; namely, because we currently have the source updating and reset and separate actions, there's technically a state which is out of sync between those two actions. This would not be directly reachable by any user interaction, but is not desirable from purely a data perspective.

He'd incorporated the block updates into the reset step in his work of #16075. The main concern I have with this is that we expand the responsibilities of a reset action (normally just a setter) to become more aware of the cause of those changes, which to me is still a two-stage process, except bundled into a single action.

epiqueras commented:
>In another of my experimental branches, I had considered whether it would be enough that the BlockEditorProvider ensure the next value it receives in a sync matches what it had expected based on what was sent. We could use this in the implementation here, then only "apply" values to blocks if they differ from what was sent by the block editor, thus triggering the BlockEditorProvider to reset its blocks based on what was applied from the editor.

That looks like it would fix this issue.

>In thinking again about the "Source API" here, and partly with regard to my previous comment at least so far as the added complexity in avoiding to mutate (and only create new references) for blocks "applied", I wonder if we could eliminate applyAll, and align what's proposed here as align closer to the update interface.

Yeah that makes more sense now that we need that equality check in the block editor provider `componentDidUpdate`.

>Considering how it might be made to be generic, the sources would need some way to subscribe to the store, and in the case of meta, receive dependent data upon whose changing should cause a re-application of sourced data on blocks.

Something along the lines of:

```js
export const reApplyOn = [ 'EDIT_POST' ]

// Also:
export const reApplyOn = { 'EDIT_POST': ( state, action ) => true || false }
```

aduth commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -983,6 +983,32 @@ export const blockListSettings = ( state = {}, action ) => {
+ *
+ * @return {[string,Object]} Updated state.
+ */
+export function lastBlockAttributesChange( state = [ null, null ], action ) {
```
> Is this exported intentionally?

It would be needed for unit tests, once I've implemented them 😄 

Despite being exported from the file, none of these reducers would be publicly available on the top-level API of the module.

epiqueras commented:
>Maybe. @youknowriad mentioned a specific concern to me that this might help address; namely, because we currently have the source updating and reset and separate actions, there's technically a state which is out of sync between those two actions. This would not be directly reachable by any user interaction, but is not desirable from purely a data perspective.

Yeah, having it in one action would be better.

aduth commented:
In speaking directly with @epiqueras , we mentioned a few ideas for revisions.

Per prior comments https://github.com/WordPress/gutenberg/pull/16402#issuecomment-508069981, https://github.com/WordPress/gutenberg/pull/16402#issuecomment-508140660, and https://github.com/WordPress/gutenberg/pull/16402#pullrequestreview-257576501, we need some way to subscribe to changes. I think this could also be used as a way to provide "shared" resources in applying attributes to the blocks, which is currently what the proposed `applyAll` was meant to address. By operating in retrieving singular block attribute values, we can also move some of the complexities of managing blocks references updates to the framework level.

Sample interface:

```js
export function* getDependencies() {}
export function* apply( attributeSchema, dependencies ) {}
export function* update( attributeSchema, value ) {}
```

Meta example:

```js
export function* getDependencies() {
	return { meta: yield select( 'core/editor', 'getEditedPostAttribute', 'meta' ) };
}

export function apply( attributeSchema, { meta } ) {
	return meta[ attributeSchema.meta ];
}

export function* update( attributeSchema, value ) {
	yield editPost( { meta: { [ attributeSchema.meta ]: value } } );
}
```

Per comments https://github.com/WordPress/gutenberg/pull/16402#discussion_r300057161 and https://github.com/WordPress/gutenberg/pull/16402#issuecomment-508200749, we may want to explore incorporating awareness of the specific block updates which contributed to `resetEditorBlocks` as part of the same action. We can move what we have today in the logic which calls to `onBlockAttributesChange` to be as part of the `onInput` and `onChange` handlers, comparing whether the last block updates have been changed. This moves much more in the direction of @youknowriad's original pull request at #16075.

Per comment https://github.com/WordPress/gutenberg/pull/16402#issuecomment-508068761, we should incorporate this fix to assure that when the BlockEditorProvider performs an "outbound sync", the value it receives next is what it assumes it should be, and reset if it is different (in the case of meta updates, we will produce a new value reference if there are multiple meta blocks which need to be updated in response to the change).

epiqueras commented:
From exploring with @youknowriad how this would work with a post-content block that uses inner blocks. We think we need a special case of the API outlined above.

`registerBlockType` settings should accept an `innerBlocksSource` property that allows you to set a custom source.

Here is what the implementation for post-content could look like:

```js
export function* getDependencies() {
    return yield select( 'core/editor', 'getEditedPostAttribute', 'content' );
}

export function apply( attributeSchema,  content ) {
    return parse(content);
}

export function* update( attributeSchema, value ) {
    yield editPost( { content: serialize(value) );
}
```

Differences to custom sources used for attributes:

- `apply`'s return value should replace the block's inner blocks.
- Editing inner blocks does not set attributes on the parent block so `update` here would never fire. We probably don't want it to fire on every edit anyways, because serializing is expensive. So, we need `update` to be called just once at the start of `savePost`.

aduth committed:
Block Editor: Skip outgoing sync next value only if matches by reference

aduth committed:
Editor: Implement editor tearDown to un-ready state

aduth committed:
Editor: Add subscription action generator for managing sources dependencies

aduth commented:
I've pushed up some revisions:

- Since #16407 was merged with specific handling for updates to meta properties, this pull request now removes that handling, as it seeks to make these unnecessary.
- Dependencies updates are now managed by a special action generator which effectively subscribes to store changes. It was done this way to keep sourcing handling colocated within the store, and to ensure that the sources dependencies getter could be implemented as a generator to make use of data controls.
   - To treat it as a proper subscription with unsubscribe, a new `tearDown` action was added, which controls the existing `isReady` state flag.
- From 5adb8ab8b62a7be9c5aa02f54f7e6b65317376bf (an alternate branch `try/editor-custom-sources`), we now ensure that the BlockEditorProvider only ignores the next value after an outbound sync if the value matches what it expected to be synced. This ensures that a full reset would occur if a meta block being updated should cause cascading effects to other meta blocks using the same property.

As can be seen in [the code](https://github.com/WordPress/gutenberg/pull/16402/files#diff-4484dcdb6f86b451bba9a7e3336bd0a2), the revisions achieve the proposed interface mentioned in my [previous comment](https://github.com/WordPress/gutenberg/pull/16402#issuecomment-508253520).

What still needs work:

- `resetEditorBlocks` is called twice because. One is caused by the `editPost` coming from a block's attribute update, the other from the BlockEditorProvider's default `onInput` / `onChange` callbacks. As mentioned in my previous comment, we should try to consolidate this to keep a single `resetEditorBlocks` action which is aware of how to call the source `update` functions and reconcile to avoid duplicate resets.
- There's a bit of wastefulness in the current implementation for how dependencies are passed to the source `apply` function. Ideally this can reuse the same cache store as in the "subscribe" behavior mentioned above.

epiqueras commented:
Nice work!

I think I have an idea that can kill both those remaining birds with one stone:

```js
export const { resetEditorBlocks, subscribeSources } = ( () => {
	const lastDependencies = new WeakMap(); // Shared cache.
	const updateDependencies = function*() {}; // Returns true if they have changed, and false if not.

	return {
		*resetEditorBlocks( blocks, options = {} ) {
			if ( aCustomSourcedAttributedHasChanged ) {
				callUpdates();

				// This makes sure `subscribeSources` doesn't call back here unless something directly updates a source.
				updateDependencies();
			}

			return {
				type: 'RESET_EDITOR_BLOCKS',
				// Reuses cache!
				blocks: yield* getBlocksWithSourcedAttributes( blocks, lastDependencies ),
				// Integrates with undo logic!
				shouldCreateUndoLevel: options.__unstableShouldCreateUndoLevel !== false,
			};
		},

		// Won't call `resetEditorBlocks` if `resetEditorBlocks` already updated dependencies.
		*subscribeSources() {},
	};
} )();
```

This approach shares the cache, makes sure we only call `resetEditorBlocks` once, and we get integration with undo logic for free.

aduth commented:
@epiqueras Good idea. It seems to illustrate: Maybe we shouldn't bother with the IIFE at all, and just create a variable(s) in the shared top-level scope?

This hints to another problem, however: Dependencies should be tracked unique per each registry. I have a few ideas for how we might incorporate this, but both add some further complexity to the solution:

1. Change our dependency tracking to account _per registry_. For example, `lastDependencies.get( source )` becomes `lastDependencies.get( registry ).get( source )`.
   - We would need some way to get a reference to the registry from within the action, likely via another control `GET_REGISTRY: createRegistryControl( ( registry ) => () => registry )`
2. Manage last dependencies in state, via a reducer + selector combination

If we hope for this sources behavior to become reusable at some point, these add a fair bit of overhead to how it would need to be implemented. At this point though, I think these goals could be optimized in future refactoring.

epiqueras commented:
Good catch!

I think we can implement your first suggestion easily by making
AWAIT_NEXT_STATE_CHANGE resolve to the registry, but your second suggestion
will scale to more use cases and fits our patterns better.

We can use the same part of the store we use for last attribute changes.

On Mon, Jul 8, 2019 at 7:11 AM Andrew Duthie <notifications@github.com>
wrote:

> @epiqueras <https://github.com/epiqueras> Good idea. It seems to
> illustrate: Maybe we shouldn't bother with the IIFE at all, and just create
> a variable(s) in the shared top-level scope?
>
> This hints to another problem, however: Dependencies should be tracked
> unique per each registry. I have a few ideas for how we might incorporate
> this, but both add some further complexity to the solution:
>
>    1. Change our dependency tracking to account *per registry*. For
>    example, lastDependencies.get( source ) becomes lastDependencies.get(
>    registry ).get( source ).
>       - We would need some way to get a reference to the registry from
>       within the action, likely via another control GET_REGISTRY:
>       createRegistryControl( ( registry ) => () => registry )
>    2. Manage last dependencies in state, via a reducer + selector
>    combination
>
> If we hope for this sources behavior to become reusable at some point,
> these add a fair bit of overhead to how it would need to be implemented. At
> this point though, I think these goals could be optimized in future
> refactoring.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/16402?email_source=notifications&email_token=AESFA2AZRSK7MPR5KWCZ3FLP6MVFPA5CNFSM4H47X6S2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZM4KHQ#issuecomment-509199646>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AESFA2A4OJ3JMQCLRT4FR5DP6MVFPANCNFSM4H47X6SQ>
> .
>

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -67,8 +122,83 @@ export function* setupEditor( post, edits, template ) {
 		blocks = synchronizeBlocksWithTemplate( blocks, template );
 	}
 
-	yield resetEditorBlocks( blocks );
```
I might run into it again in the future; I don't recall why I needed to reorder `resetEditorBlocks`, but it causes an issue where a new post will prompt about unsaved changes when `SCRIPT_DEBUG` is `true`, due to a fragile guarantee the current order establishes that causes "dirtiness" state to be unset by the `setupEditorState` action. This incidentally resolves the fact we dispatch this `setupEditor` action twice in an editing session, since it [happens in `constructor`](https://github.com/WordPress/gutenberg/blob/0d9efe27209c396ec6209622d58d706f2e8167c5/packages/editor/src/components/provider/index.js#L57) (rather than `componentDidMount`) and React helpfully detects side-effects via [double-invoking intended non-side-effect lifecycle functions](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects).

aduth replied:
Oh, I remember now: We want the sourced attributes to be applied in `resetEditorBlocks`, which requires that we set the post state first (so that post meta can be read).

In that case, we might need to address the underlying issue with the `EditorProvider` lifecycle, or endure the prompts in `SCRIPT_DEBUG` (preferably not).

epiqueras replied:
From our conversation:

>So `resetEditorBlocks` sets the flag to true if the blocks are different.
`setupEditorState` sets it to false again.

>Basically it only worked because `setupEditorState` was called as last, so it reset the dirty flag

>The dirtying was actually pretty simple if we leave things as they are, and just call `resetPost` before `resetEditorBlocks` (so that meta values are available for applying)

aduth replied:
At some point, we should also look to refactor this so that the editor module doesn't need to store a copy of the "current" post, but instead calls to the `@wordpress/core-data` module to apply the edits on the canonical post object.

```js
wp.data.select( 'core' ).getEntityRecord( 'postType', 'post', wp.data.select( 'core/editor' ).getCurrentPostId() );
```

aduth committed:
Editor: Track source dependencies by registry

aduth committed:
Editor: Prepare initial blocks reset for applied post-sourced values

epiqueras commented:
`getDependencies` should also take the schema so we can do things like:

```js
export function* getDependencies( schema ) {
	return {
		meta: yield select( 'core/editor', 'getEditedPostAttribute', schema.postAttribute ),
	};
}
```

For a `postAttribute` source.

epiqueras commented:
Even better:

```js
export function* getDependencies( schema ) {
	return _.get(yield select( 'core/editor', 'getEditedPostAttribute', schema.attribute ), schema.property );
}
```

For a title source:

```json
"attributes": {
		"title": {
			"type": "string",
			"source": "post",
			"attribute": "title"
		}
	}
```

For meta:

```json
"attributes": {
		"something": {
			"type": "string",
			"source": "post",
			"attribute": "meta",
			"property": "something"
		}
	}
```

aduth commented:
> `getDependencies` should also take the schema so we can do things like:

The problem with this is that it can be distinct by block type, rather than strictly for all blocks / block types of a given source. The specific meta property from which attribute values are sourced can vary by block type.

epiqueras commented:
>The problem with this is that it can be distinct by block type, rather than strictly for all blocks / block types of a given source. The specific meta property from which attribute values are sourced can vary by block type.

You are right, I missed that.

We would need to loop over registered block attribute schemas and use their serializations as cache keys for this to work. That can quickly get out of hand.

In any case, we can still use the property pattern in `apply` and `update` so that the `post` source can be used for meta. We edit post attributes and meta through the same action so there is no performance difference.

aduth committed:
Editor: Yield editor setup action after state preparation

Ensure initial edits set to expected values, since preceding resetPost would unset them otherwise if already assigned

epiqueras commented:
Like this:

```js
export function* getDependencies() {
	return {
		post: yield select( 'core/editor', 'getCurrentPost' ),
		content: yield select( 'core/editor', 'getEditedPostContent' ),
		edits: yield select( 'core/editor', 'getPostEdits' ),
	};
}

export function apply( { attribute }, { post, content, edits } ) {
	if ( attribute === 'content' ) {
		return content;
	}

	if ( undefined !== get( edits, attribute ) ) {
		return get( edits, attribute );
	}

	if ( EDIT_MERGE_PROPERTIES.has( attribute ) ) {
		return {
			...get( post, attribute ),
			...get( edits, attribute ),
		};
	}

	return get( post, attribute );
}

export function* update( { attribute }, value ) {
	yield editPost( set( {}, attribute, value ) );
}
```

epiqueras commented:
Small issue:

`updateBlockSources` does not work with inner blocks, because `getEditorBlocks` does not return a flat list.

aduth committed:
Editor: Update block sources as part of reset step

aduth commented:
> Small issue: `updateBlockSources` does not work with inner blocks, because `getEditorBlocks` does not return a flat list.

I pushed some updates which removes `updateBlockSources` altogether, though I suspect the problem still remains. We might need to do some recursion / stack accumulation when encountering inner blocks to apply/update down through the hierarchy.

aduth committed:
Editor: Retrieve registry at most once in applying sourced attributes

aduth committed:
Editor: Account for source values apply / update in inner blocks

aduth commented:
>We might need to do some recursion / stack accumulation when encountering inner blocks to apply/update down through the hierarchy.

This worked pretty seamlessly. More complexity / overhead, but it gets the job done.

![meta-inner-blocks](https://user-images.githubusercontent.com/1779930/60837062-3f8e4c00-a195-11e9-87eb-6024a3be92dd.gif)

aduth committed:
Editor: Update block sources documentation per interface changes

aduth committed:
Block Editor: Add lastBlockAttributesChange unit tests

aduth committed:
Block Editor: Document, test getLastBlockAttributesChange selector

epiqueras commented:
Yeah that did it! I'll give this another review.

epiqueras reviewed (commented):
I think this is almost ready to merge 🎉 

Just a couple more things and some typos to fix.

epiqueras commented:
I'm using this in https://github.com/WordPress/gutenberg/tree/try/expanding-the-editor-outside-post-content and it's working nicely! The API is really intuitive, integrates well with the codebase, and maintains backwards compatibility.

aduth commented on packages/block-editor/src/components/provider/index.js:
```diff
@@ -68,31 +68,47 @@ class BlockEditorProvider extends Component {
 				return;
 			}
 
+			if ( newLastBlockAttributesChange !== lastBlockAttributesChange ) {
```
If we did something like this, I think it would have to be a tuple of the entire set of attributes changes, as otherwise I'm not sure when we would ever "clean up" old attributes changes? And more work to iterate through the changes.

Something like:

```
[ lastAttributesChanges, newAttributesChanges ]
```

... where each has the shape:

```
{ [ clientId ]: { [ attribute ]: change } }
```

aduth replied:
Another option: In @youknowriad's #16075, he [implemented this reducer](https://github.com/WordPress/gutenberg/pull/16075/files#diff-ff2edcc560dbdf06e4a49898f00e323aR986) to always reset to `null` when the action dispatched is anything other than a block attributes update. That way we could infer that a change has occurred if the value resulting from `getLastBlockAttributesChange` is anything other than `null`.

The only thing [unclear to me](https://github.com/WordPress/gutenberg/pull/16075/files#r299628880) is whether we can safely assume that the value will be non-`null` as long as we need it to be.

aduth replied:
> Another option: In @youknowriad's #16075, he [implemented this reducer](https://github.com/WordPress/gutenberg/pull/16075/files#diff-ff2edcc560dbdf06e4a49898f00e323aR986) to always reset to `null` when the action dispatched is anything other than a block attributes update. That way we could infer that a change has occurred if the value resulting from `getLastBlockAttributesChange` is anything other than `null`.

I implemented this in 085c9eb64. It seems to work quite well, and gains all of the benefits you were hoping from your suggestion.

epiqueras replied:
>as otherwise I'm not sure when we would ever "clean up" old attributes changes?

In hindsight, we would have had this problem with any data structure. It arises from the fact that you want to "cleanup" after checking for a change and reacting on it. This can't be done easily if the data is shared like in my suggestion. You would need to keep local flags everywhere you use it.

>I implemented this in 085c9eb. It seems to work quite well, and gains all of the benefits you were hoping from your suggestion as far as simplifying the implementation to avoid onBlockAttributesUpdate and instead fetch the updates directly within resetEditorBlocks.

This is perfect for what we need now. Much cleaner 👍

aduth commented on packages/block-editor/src/components/provider/index.js:
```diff
@@ -34,10 +34,10 @@ class BlockEditorProvider extends Component {
 
-		if ( this.isSyncingOutcomingValue ) {
-			this.isSyncingOutcomingValue = false;
+		if ( this.isSyncingOutcomingValue === value ) {
```
> Because we don't update `this.isSyncingOutcomingValue` when the new value is different, an update where the value is set to the last outcoming value or `null` would be ignored. Certain implementations of "undo" could be affected by this (maybe ours?).

That's a good point. Should only the addition of `this.isSyncingOutcomingValue = null;` to the `else if` condition be sufficient though? Is testing for `!== null` necessary in the first condition?

aduth replied:
I wonder if we could just form it like this, where assigning to `null` is fine to _always_ occur in `componentDidUpdate`, after the logic to consider whether the value has changed, is not a result of an outgoing sync.

```js
componentDidUpdate( prevProps ) {
	// ...

	if ( value !== prevProps.value && this.isSyncingOutcomingValue !== value ) {
		this.isSyncingIncomingValue = value;
		resetBlocks( value );
	}

	this.isSyncingOutcomingValue = null;
}
```

epiqueras replied:
>Should only the addition of this.isSyncingOutcomingValue = null; to the else if condition be sufficient though? 

I think so.

>Is testing for !== null necessary in the first condition?

Yeah, otherwise, the parent setting blocks to `null` would be ignored if there is no pending outcoming value.

>I wonder if we could just form it like this, where assigning to null is fine to always occur in componentDidUpdate, after the logic to consider whether the value has changed, is not a result of an outgoing sync.

This will force a double reset if there is any update scheduled by React in-between when an outcoming sync starts and ends.

aduth replied:
> This will force a double reset if there is any update scheduled by React in-between when an outcoming sync starts and ends.

That's fair. I'm not sure we expect that to happen, but it's probably worth the precaution. I'll revise accordingly.

epiqueras replied:
Yeah, before, you could get a double reset either from the parent modifying the value or from there being an update in-between that cleared `isSyncingOutcomingValue`.

React updates can get pretty crazy, even if they can be deterministic at times, the order is so hard to wrap your head around that we are better off just being defensive.

aduth committed:
Editor: Fix block sources getDependencies grammar

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

aduth committed:
Editor: Fix block sources meta getDependencies JSDoc typo

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

aduth commented on packages/editor/src/store/block-sources/meta.js:
```diff
@@ -0,0 +1,54 @@
+ * equality of the returned object), blocks are reset to apply the new sourced
+ * value.
+ *
+ * @yield {Object} Optional yielded controls.
```
This makes sense if we consider alignment within groups. The only thing I wonder is whether we should consider `yield` and `return` as a group. There's no [standard](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions) around this. They seem of the same "sort", which makes me inclined to think this arrangement (and your suggestion) are the ideal one.

epiqueras replied:
Consensus in the core-js chat was to separate them into two different groups.

aduth committed:
Editor: Align block sources getDependencies JSDoc tags

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

aduth committed:
Editor: Resolve awaitNextStateChange JSDoc typo

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

aduth committed:
Editor: Resolve getRegistry JSDoc typo

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

aduth committed:
Block Editor: Always unset expected outbound sync value in provider update

aduth committed:
Editor: Detect block attributes change as direct result of last action

Infer change as non-null state from lastBlockAttributesChange

aduth committed:
Documentation: Regenerate editor, block-editor data documentation

aduth committed:
Block Editor: Unset outbound sync value in editor provider reset condition

aduth committed:
Editor: Update getDependencies JSDoc to separate yield grouping

aduth committed:
Editor: Dispatch SETUP_EDITOR prior to blocks reset

SETUP_EDITOR is responsible for setting the initial edits, which are expected to be unset by a subsequent block reset

aduth committed:
Editor: Update actions tests per reordered SETUP_EDITOR

youknowriad commented on packages/block-editor/src/components/provider/index.js:
```diff
@@ -34,10 +34,21 @@ class BlockEditorProvider extends Component {
+			// bound sync, unset the outbound value to avoid considering it in
+			// subsequent renders.
+			this.isSyncingOutcomingValue = null;
+			this.isSyncingIncomingValue = value;
```
I know these changes are really tricky and fragile. The good news is that we have e2e tests for most of the use-cases and you can endup fixing one but breaking others :P

youknowriad commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -195,29 +195,6 @@ export function isUpdatingSameBlockAttribute( action, lastAction ) {
- *
- * @return {Function} Enhanced reducer function.
- */
-const withPostMetaUpdateCacheReset = ( reducer ) => ( state, action ) => {
```
❤️

youknowriad commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1228,17 +1205,44 @@ export const blockListSettings = ( state = {}, action ) => {
+
+		case 'UPDATE_BLOCK_ATTRIBUTES':
+			return { [ action.clientId ]: action.attributes };
+	}
```
What happens for other actions that change multiple blocks, like RESET_BLOCKS... (Thinking about a third-party plugin triggering that action and not the block editor provider), should we try to list the changes or do we just trigger a full sync? What happens now?

aduth replied:
> What happens for other actions that change multiple blocks, like RESET_BLOCKS... (Thinking about a third-party plugin triggering that action and not the block editor provider), should we try to list the changes or do we just trigger a full sync? What happens now?

It wouldn't try to sync meta attributes from the updates, which is the same as how it behaves today in master. I think it might be a bit difficult to infer from the reset whether those meta values should be considered canonical, or whether the source should be applied again by the editor. The latter is what I expect happens in the current state of this branch.

youknowriad commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1421,19 +1383,16 @@ export function __unstableIsLastBlockChangeIgnored( state ) {
-	}
-
-	return get( state, [ 'settings', '__experimentalMetaSource', 'value', key ] );
+export function getLastBlockAttributesChange( state ) {
```
How confident are we about the long term validity of this selector? should it be experimental for some time? (I guess same question for the two actions being introduced)

aduth replied:
> How confident are we about the long term validity of this selector? should it be experimental for some time? (I guess same question for the two actions being introduced)

I think it would always almost always for internal use, but I suppose that doesn't help us much as far as backwards-compatibility commitment. I don't mind to mark them as experimental for a few releases.

I don't know if we might consider another sort of unstable naming convention for `__internal` methods, which this would fit. There is some precedent with WordPress and its underscore-prefixed functions (e.g. [`_wp_admin_bar_init`](https://developer.wordpress.org/reference/functions/_wp_admin_bar_init/)).

aduth replied:
> How confident are we about the long term validity of this selector? should it be experimental for some time? (I guess same question for the two actions being introduced)

I'm going to change these to be experimental prefixed.

As an aside, I also questioned how this should be named considering pluralization, since there can be one or multiple blocks which have one or multiple attributes changed. Having it as experimental would let us rename if we settle on a better form (`getLastBlocksAttributesChanges`? `getLastBlocksAttributeChanges`? `getLastBlockAttributeChanges`?)

epiqueras replied:
>`getLastBlockAttributeChanges`

It returns changes to attributes.

aduth replied:
> > `getLastBlockAttributeChanges`
> 
> It returns changes to attributes.

Works for me. Updated in d955e89.

youknowriad commented on packages/editor/src/store/actions.js:
```diff
@@ -730,10 +904,47 @@ export function unlockPostSaving( lockName ) {
+		blocks: yield* getBlocksWithSourcedAttributes( blocks ),
 		shouldCreateUndoLevel: options.__unstableShouldCreateUndoLevel !== false,
 	};
 }
```
To be completely fair, I got the idea of most of the functions here but it's hard to follow though. a schema or something somewhere might be good for maintenance.

aduth replied:
> To be completely fair, I got the idea of most of the functions here but it's hard to follow though. a schema or something somewhere might be good for maintenance.

I'd also wondered if I should try to do a better job here of extracting behaviors into functions, maybe also moving some of them into `block-sources/index.js`. Some of this could be future refactors as well.

youknowriad commented on packages/editor/src/store/block-sources/README.md:
```diff
@@ -0,0 +1,22 @@
+
+## Source API
+
+### `getDependencies`
```
Makes me wonder if we could be more granular at some point. Say update only blocks that depend on a specific meta key if that meta change, otherwise don't update any block.

I guess it would complicate things as we need to track blocks per specific key. 

Maybe a future optimization if needed.

aduth replied:
> Makes me wonder if we could be more granular at some point. Say update only blocks that depend on a specific meta key if that meta change, otherwise don't update any block.

@epiqueras raised a very similar point for block types which depend on properties from a post, since the way those would likely need to be implemented today is to depend on the full post object, since it's only at the point of `apply` that we can determine the relevant property from the schema (since it varies by block type).

I agree we could probably explore this as a future optimization.

youknowriad reviewed (approved):
This is looking good for me.

youknowriad commented:
We should run performance tests against this PR.

aduth commented:
> We should run performance tests against this PR.

Hmm, it does seem to regress a bit.

Measure|Before|After
---|---|--
Average time to load:|11067ms|15961ms
Average time to DOM content load:|10379ms|15130ms
Average time to type character:|85.375ms|102.335ms
Slowest time to type character:|213ms|237ms
Fastest time to type character:|60ms|72ms

I expected some give and take. The refactorings to block editor (specifically selectors attributes merging and reducer enhancers) seemed like they could bring benefits. On the other hand, we now have a new registry subscription for `getDependencies`. I also think there's probably quite a bit of overhead now in `resetEditorBlocks` which wasn't there previously (typing a character performs two iterations over blocks, one to `update` sources and a second to `apply`). I have a feeling those could probably be consolidated into a single iteration, or further that we be more strategic about how we apply block updates to try to limit how we iterate (e.g. know which sources need to be updated without iterating all blocks, know which blocks updated sources need to be applied upon).

aduth commented:
The current end-to-end test failure occurs here:

https://github.com/WordPress/gutenberg/blob/385d79c387581abb5fcdf4d8989a39c5cb478f18/packages/editor/src/store/actions.js#L64

Specifically, `wp.blocks.getBlockType( 'core/nextpage' ).attributes` is `undefined`.

It seems to me we should have some stronger guarantees about the shape that a block type returns, i.e. even if a block type is registered without an `attributes` schema, the value is defaulted to an empty object.

aduth committed:
Block API: Stub default attributes, keywords values for block type registration

aduth committed:
Block API: Update documentation per formation of WPBlockTypeIconDescriptor typedef

aduth committed:
Editor: Iterate last block changes in updating sources

aduth commented:
The changes in 54c5b8a appear to reclaim all (and more?) of the lost performance.

>Average time to load: 5804ms
>Average time to DOM content load: 5139ms
>Average time to type character: 83.535ms
>Slowest time to type character: 211ms
>Fastest time to type character: 58ms

The main difference is in how updates are synced from block updates:

- Avoid creating a clone of the reset blocks
- Avoid needing to recurse into the reset blocks inner blocks
- Avoid needing to iterate the reset blocks at all, instead iterating only the _updates_
- Avoid considering a block type which has already been considered
- Avoid considering a source which has already been considered (we may have had duplicate `update` calls before this)

aduth committed:
Editor: Update source even if already updated for block updates

A block may have multiple attributes which independently source uniquely from the same source type (e.g. two different meta properties)

epiqueras commented on packages/editor/src/store/actions.js:
```diff
@@ -910,28 +910,27 @@ export function* resetEditorBlocks( blocks, options = {} ) {
+			updatedBlockTypes.add( blockName );
+			const blockType = yield select( 'core/blocks', 'getBlockType', blockName );
 
 			for ( const [ attributeName, schema ] of Object.entries( blockType.attributes ) ) {
```
Nice job on this optimization!

It would also be more intuitive to just loop over `attributes` here.

aduth replied:
> It would also be more intuitive to just loop over `attributes` here.

Agreed! This is another optimization, since the updated attributes are likely a subset of the attributes of a block type (and most likely just a single attribute), so we can avoid iterating over all attributes of a block type by arranging it this way.

I've revised this in 38d27e1.

aduth committed:
Editor: Iterate updated attributes from block reset source update

Optimization since the attributes are likely a subset (most likely a single attribute), so we can avoid iterating all attributes defined in a block type.

aduth committed:
Editor: Mark custom sources selectors, actions as experimental

aduth committed:
Editor: Remove redundant Object#hasOwnProperty in iterating attributes

aduth committed:
Block Editor: Rename getLastBlockAttributesChange to getLastBlockAttributeChanges

aduth committed:
Editor: Use getRegistry control for next change subscription

aduth commented:
I've rebased to resolve conflicts introduced by #16184. I started squashing a few commits to clean up the history, but it got a bit unwieldy, so I left it more-or-less as it was (shouldn't matter much anyways since I'll Squash and Merge). I added a new commit 468f908 which removes an earlier approach to retrieving the registry in the resolution of `subscribeSources` `awaitNextStateChange`, since there was a later addition of a `getRegistry` control anyways.

I'm planning to merge this shortly.

As follow-up tasks considered in this pull request:

- Resolve blocks validation issues with custom sources (#4989). This pull request does not improve this situation. I anticipate the resolution here involves deferring blocks validation until after a first pass to `apply` source values.
- Mentioned in passing at https://github.com/WordPress/gutenberg/pull/16402#discussion_r301216604, there's a potential refactor with `getCurrentPost`. It's not very relevant for custom sources, but worth exploring separately. I've created a new issue at #16520 to track this.
- In iterations, consider reorganization of the functions / logic to better isolate the complexity
- Start exploring other custom sources (e.g. post properties in #16485)

aduth commented:
Thanks @epiqueras and @youknowriad for your detailed feedback and suggestions here!

mcsf commented on packages/editor/src/store/actions.js:
```diff
@@ -32,6 +33,120 @@ import {
+
+			if ( ! lastBlockSourceDependenciesByRegistry.has( registry ) ) {
+				continue;
+			}
```
Is the `if` condition expected to evaluate differently across `for` iterations? If not, we could place it as a condition for stepping into the `for` loop.

mcsf commented on packages/editor/src/store/actions.js:
```diff
@@ -67,8 +175,76 @@ export function* setupEditor( post, edits, template ) {
+				lastBlockSourceDependenciesByRegistry.set( registry, new WeakMap );
+			}
+
+			const lastBlockSourceDependencies = lastBlockSourceDependenciesByRegistry.get( registry );
```
Can probably factor out this assignment and the optional `WeakMap` initialisation that comes before it.

mcsf reviewed (commented):
This is really great work that I've only now caught up with. Left some minor notes.

aduth commented:
Thanks for reviewing, @mcsf! Both are valid points. See #16839.
